From 0bc6387abe8e8feac8fac1cc6c54cebe59cd37ef Mon Sep 17 00:00:00 2001
From: Sharaf Zaman <sharafzaz121@gmail.com>
Date: Tue, 10 Nov 2020 18:24:32 +0000
Subject: [PATCH] Android: Handle the ACTION_CANCEL Event from Android

Prior to this actions from weren't handled explicity, this made events
after the ACTION_CANCEL was sent to continue from previous state, rather
than starting over.
---
 .../src/org/qtproject/qt5/android/QtNative.java    | 14 +-------------
 src/plugins/platforms/android/androidjniinput.cpp  | 14 ++++++++++++--
 2 files changed, 13 insertions(+), 15 deletions(-)

diff --git a/src/android/jar/src/org/qtproject/qt5/android/QtNative.java b/src/android/jar/src/org/qtproject/qt5/android/QtNative.java
index 0ec8170321..0102157c98 100644
--- a/src/android/jar/src/org/qtproject/qt5/android/QtNative.java
+++ b/src/android/jar/src/org/qtproject/qt5/android/QtNative.java
@@ -711,19 +711,7 @@ public class QtNative
                              event.getOrientation(i),
                              event.getPressure(i));
             }
-
-            switch (event.getAction()) {
-                case MotionEvent.ACTION_DOWN:
-                    touchEnd(id, 0);
-                    break;
-
-                case MotionEvent.ACTION_UP:
-                    touchEnd(id, 2);
-                    break;
-
-                default:
-                    touchEnd(id, 1);
-            }
+            touchEnd(id, event.getAction());
         }
     }
 
diff --git a/src/plugins/platforms/android/androidjniinput.cpp b/src/plugins/platforms/android/androidjniinput.cpp
index e287f4121f..33778e56df 100644
--- a/src/plugins/platforms/android/androidjniinput.cpp
+++ b/src/plugins/platforms/android/androidjniinput.cpp
@@ -38,6 +38,8 @@
 **
 ****************************************************************************/
 
+#include <android/input.h>
+
 #include <QtGui/qtguiglobal.h>
 
 #include "androidjniinput.h"
@@ -301,7 +303,7 @@ namespace QtAndroidInput
         }
     }
 
-    static void touchEnd(JNIEnv */*env*/, jobject /*thiz*/, jint /*winId*/, jint /*action*/)
+    static void touchEnd(JNIEnv */*env*/, jobject /*thiz*/, jint /*winId*/, jint action)
     {
         if (m_touchPoints.isEmpty())
             return;
@@ -324,7 +326,15 @@ namespace QtAndroidInput
         }
 
         QWindow *window = QtAndroid::topLevelWindowAt(m_touchPoints.at(0).area.center().toPoint());
-        QWindowSystemInterface::handleTouchEvent(window, touchDevice, m_touchPoints);
+
+        switch (action) {
+        case AMOTION_EVENT_ACTION_CANCEL:
+            QWindowSystemInterface::handleTouchCancelEvent(window, touchDevice);
+            break;
+        default:
+            QWindowSystemInterface::handleTouchEvent(window, touchDevice, m_touchPoints);
+            break;
+        }
     }
 
     static bool isTabletEventSupported(JNIEnv */*env*/, jobject /*thiz*/)
-- 
2.28.0

